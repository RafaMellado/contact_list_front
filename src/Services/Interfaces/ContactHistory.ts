export interface ContactHistoryDTO {
  id: number;
  givenname: string;
  surname: string;
  email: string;
  phone: string;
  contact_id: number;
  created_at: string;
  updated_at: string;
}

export interface ContactHistory {
  givenname: string;
  surname: string;
  email: string;
  phone: string;
  created_at: string;
}
