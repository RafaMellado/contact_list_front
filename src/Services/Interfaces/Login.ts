export interface LoginRequestBody {
  email: string;
  password: string;
}

export interface LoginResponse {
  token: string;
  username: string;
  exp: string;
}

export interface LoginError {
  error: string;
}
