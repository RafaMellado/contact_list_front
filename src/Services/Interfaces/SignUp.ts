export interface SignUpRequestBody {
  username: string;
  email: string;
  password: string;
  password_confirmation: string;
}

export interface SignUpResponse {
  username: string;
  email: string;
}
