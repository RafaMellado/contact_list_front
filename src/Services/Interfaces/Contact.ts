import { ContactHistory, ContactHistoryDTO } from "./ContactHistory";

export interface ContactDTO {
  id: number;
  givenname: string;
  surname: string;
  email: string;
  phone: string;
  contact_book_id: number;
  created_at: string;
  updated_at: string;
}

export interface ContactWithContactHistoriesDTO {
  id: number;
  givenname: string;
  surname: string;
  email: string;
  phone: string;
  contact_book_id: number;
  contact_histories: Array<ContactHistoryDTO>;
  created_at: string;
  updated_at: string;
}

export interface Contact {
  id: number;
  givenname: string;
  surname: string;
  email: string;
  phone: string;
  contact_book_id: number;
}

export interface ContactWithContactHistories {
  id: number;
  givenname: string;
  surname: string;
  email: string;
  phone: string;
  contact_book_id: number;
  contact_histories: Array<ContactHistory>;
}

export interface ContactRequestBody {
  givenname: string;
  surname: string;
  email: string;
  phone: string;
  contact_book_id: number;
}

export interface ContactRequestError {
  errors: {
    email: Array<{ error: string; value: string }>;
  };
}
