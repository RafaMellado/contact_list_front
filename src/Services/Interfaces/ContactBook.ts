import { Contact } from "./Contact";

export interface ContactBookDTO {
  id: number;
  name: string;
  user_id: number;
  created_at: string;
  updated_at: string;
}

export interface ContactBookWithContactsDTO {
  id: number;
  name: string;
  user_id: number;
  contacts: Array<Contact>;
  created_at: string;
  updated_at: string;
}

export interface ContactBook {
  id: number;
  name: string;
  user_id: number;
}

export interface ContactBookWithContacts {
  id: number;
  name: string;
  user_id: number;
  contacts: Array<Contact>;
}

export interface ContactBookRequestBody {
  name: string;
}

export interface ContactBookRequestError {
  errors: {
    name: Array<{ error: string; value: string }>;
  };
}
