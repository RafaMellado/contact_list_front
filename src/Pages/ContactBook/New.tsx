import { FormEvent, useRef } from "react";
import { Button, Form } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router-dom";
import ContactBookService from "../../Services/ContactBookService";

export function New() {
  const navigate = useNavigate();
  const { t } = useTranslation();

  const name = useRef<HTMLInputElement>(null);

  const TRANSLATIONS: string = "contactBookNew";

  const addContactBook = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    ContactBookService.new({ name: String(name?.current?.value) });

    navigate("/home");
  };

  return (
    <>
      <Form onSubmit={addContactBook}>
        <Form.Group className="mb-3">
          <Form.Label>{t<string>(`${TRANSLATIONS}.name`)}</Form.Label>
          <Form.Control
            as="input"
            type="text"
            placeholder={t<string>(`${TRANSLATIONS}.name`)}
            maxLength={20}
            minLength={4}
            required
            ref={name}
          />
        </Form.Group>

        <Button variant="primary" type="submit">
          {t<string>(`${TRANSLATIONS}.submit`)}
        </Button>
      </Form>
    </>
  );
}
