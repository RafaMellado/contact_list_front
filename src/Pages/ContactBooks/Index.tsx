import { useEffect, useState } from "react";
import { Button, Col, Row } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router-dom";
import { ContactBook } from "../../Services/Interfaces/ContactBook";
import ContactBooksService from "../../Services/ContactBooksService";
import { ContactBookCard } from "../../Components/ContactBookCard";

export function ContactBookIndex() {
  const [contactBooks, setContactBooks] = useState<ContactBook[]>([]);
  const navigate = useNavigate();
  const { t } = useTranslation();

  const TRANSLATIONS: string = "contactBooksIndex";

  useEffect(() => {
    getItems();
  }, []);

  const getItems = async (params = {}) => {
    const response = await ContactBooksService.index(params);

    setContactBooks(response);
  };

  const addContactBook = () => {
    navigate("/contact-book/new");
  };

  const editContactBook = (id: number) => {
    navigate(`/contact-book/${id}/edit`);
  };

  const goToContactBook = (id: number) => {
    navigate(`/contact-book/show/${id}`);
  };

  const deleteContactBook = async (id: number) => {
    await ContactBooksService.delete(id);

    reassignContactBooks(id);
  };

  const reassignContactBooks = (deletedId: number) => {
    setContactBooks(contactBooks.filter((item) => item.id !== deletedId));
  };

  return (
    <Row>
      <div className="d-flex justify-content-between align-items-center flex-column flex-md-row mb-5">
        <h2>{t<string>(`${TRANSLATIONS}.title`)}</h2>

        <Button
          className="mt-md-0 mt-2"
          data-testid="contact-book-add-btn"
          onClick={addContactBook}
        >
          {t<string>(`${TRANSLATIONS}.addOther`)}
        </Button>
      </div>

      <div className="mb-4">
        <input
          placeholder={t<string>(`${TRANSLATIONS}.search`)}
          onChange={(event) =>
            getItems({ filter: { name: event.target.value } })
          }
        ></input>
      </div>

      {contactBooks.map((item: ContactBook) => {
        const { id, name } = item;

        return (
          <Col
            xs={12}
            sm={6}
            md={3}
            key={id}
            onClick={() => goToContactBook(id)}
            data-testid={`contact-book-${id}`}
            className="mb-4"
          >
            <ContactBookCard
              name={name}
              editFn={() => editContactBook(id)}
              deleteFn={() => deleteContactBook(id)}
            />
          </Col>
        );
      })}
    </Row>
  );
}
