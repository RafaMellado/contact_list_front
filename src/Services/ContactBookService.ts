import { ContactBook, ContactBookRequestBody } from "./Interfaces/ContactBook";
import ContactBookRepository from "./Repository/ContactBookRepository";

class ContactBookService {
  index(): Promise<ContactBook[]> {
    return ContactBookRepository.index();
  }

  new(data: ContactBookRequestBody): Promise<ContactBook> {
    return ContactBookRepository.post(data);
  }
}

export default new ContactBookService();
