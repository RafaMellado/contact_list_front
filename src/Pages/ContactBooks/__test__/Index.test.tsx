import { ContactBookIndex } from "../Index";
import { createMemoryHistory } from "history";
import {
  act,
  fireEvent,
  render,
  RenderResult,
  waitFor,
} from "@testing-library/react";
import { Router } from "react-router-dom";
import { useState as useStateMock } from "react";
import ContactBookService from "../../../Services/ContactBooksService";
import { ContactBook } from "../../../Services/Interfaces/ContactBook";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock("../../../Services/ContactBooksService", () => ({
  index: () => [{ id: 1, name: "test", user_id: 1 }] as ContactBook[],
  delete: jest.fn(),
}));

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe(`<${ContactBookIndex.name} />`, () => {
  const setState = jest.fn();

  beforeEach(() => {
    (useStateMock as jest.Mock).mockImplementation(() => [
      [{ id: 1, name: "test", user_id: 1 }],
      setState,
    ]);
  });

  const history = createMemoryHistory();
  const factoryComponent = () =>
    render(
      <Router location="/" navigator={history}>
        <ContactBookIndex />
      </Router>
    );

  test("renders correctly the component", () => {
    let component!: RenderResult;

    act(() => {
      component = factoryComponent();

      waitFor(() => {
        act(() => {
          expect(ContactBookService.index).toHaveBeenCalled();
        });
      });
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  test("add btn redirects to add contact book page", () => {
    let component!: RenderResult;

    act(() => {
      component = factoryComponent();

      waitFor(() => {
        act(() => {
          expect(ContactBookService.index).toHaveBeenCalled();
        });
      });
    });

    fireEvent.click(component.getByTestId("contact-book-add-btn"));

    expect(history.location.pathname).toBe("/contact-book/new");
  });

  test("click on card and go to show", () => {
    let component!: RenderResult;

    act(() => {
      component = factoryComponent();

      waitFor(() => {
        act(() => {
          expect(ContactBookService.index).toHaveBeenCalled();
        });
      });
    });

    fireEvent.click(component.getByTestId("contact-book-1"));

    expect(history.location.pathname).toBe("/contact-book/show/1");
  });

  test("click on card edit icon and go to show", () => {
    let component!: RenderResult;

    act(() => {
      component = factoryComponent();

      waitFor(() => {
        act(() => {
          expect(ContactBookService.index).toHaveBeenCalled();
        });
      });
    });

    fireEvent.click(component.getByTestId("contact-book-card-edit-icon"));

    expect(history.location.pathname).toBe("/contact-book/1/edit");
  });

  test("click on card delete icon and call delete", () => {
    let component!: RenderResult;

    act(() => {
      component = factoryComponent();

      waitFor(() => {
        act(() => {
          expect(ContactBookService.index).toHaveBeenCalled();
        });
      });
    });

    fireEvent.click(component.getByTestId("contact-book-card-delete-icon"));

    expect(ContactBookService.delete).toHaveBeenCalled();
  });
});
