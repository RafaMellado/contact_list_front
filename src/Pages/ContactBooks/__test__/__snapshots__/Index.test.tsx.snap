// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactBookIndex /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="d-flex justify-content-between align-items-center flex-column flex-md-row mb-5"
    >
      <h2>
        contactBooksIndex.title
      </h2>
      <button
        class="mt-md-0 mt-2 btn btn-primary"
        data-testid="contact-book-add-btn"
        type="button"
      >
        contactBooksIndex.addOther
      </button>
    </div>
    <div
      class="mb-4"
    >
      <input
        placeholder="contactBooksIndex.search"
      />
    </div>
    <div
      class="mb-4 col-md-3 col-sm-6 col-12"
      data-testid="contact-book-1"
    >
      <div
        class="platform-card card"
      >
        <div
          class="d-flex justify-content-between flex-column card-body"
        >
          <div
            class="d-flex"
          >
            <svg
              class="mt-1 me-2 platform-card-book-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 1h6v7a.5.5 0 0 1-.757.429L9 7.083 6.757 8.43A.5.5 0 0 1 6 8V1z"
                fill-rule="evenodd"
              />
              <path
                d="M3 0h10a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H3a2 2 0 0 1-2-2v-1h1v1a1 1 0 0 0 1 1h10a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H3a1 1 0 0 0-1 1v1H1V2a2 2 0 0 1 2-2z"
              />
              <path
                d="M1 5v-.5a.5.5 0 0 1 1 0V5h.5a.5.5 0 0 1 0 1h-2a.5.5 0 0 1 0-1H1zm0 3v-.5a.5.5 0 0 1 1 0V8h.5a.5.5 0 0 1 0 1h-2a.5.5 0 0 1 0-1H1zm0 3v-.5a.5.5 0 0 1 1 0v.5h.5a.5.5 0 0 1 0 1h-2a.5.5 0 0 1 0-1H1z"
              />
            </svg>
            <h5
              class="platform-card-text"
            >
              test
            </h5>
          </div>
          <div
            class="d-flex justify-content-end mt-2 platform-card-icons "
          >
            <svg
              class="platform-card-icon"
              data-testid="contact-book-card-edit-icon"
              fill="currentColor"
              height="1em"
              role="button"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.502 1.94a.5.5 0 0 1 0 .706L14.459 3.69l-2-2L13.502.646a.5.5 0 0 1 .707 0l1.293 1.293zm-1.75 2.456-2-2L4.939 9.21a.5.5 0 0 0-.121.196l-.805 2.414a.25.25 0 0 0 .316.316l2.414-.805a.5.5 0 0 0 .196-.12l6.813-6.814z"
              />
              <path
                d="M1 13.5A1.5 1.5 0 0 0 2.5 15h11a1.5 1.5 0 0 0 1.5-1.5v-6a.5.5 0 0 0-1 0v6a.5.5 0 0 1-.5.5h-11a.5.5 0 0 1-.5-.5v-11a.5.5 0 0 1 .5-.5H9a.5.5 0 0 0 0-1H2.5A1.5 1.5 0 0 0 1 2.5v11z"
                fill-rule="evenodd"
              />
            </svg>
            <svg
              class="ms-2 platform-card-icon"
              data-testid="contact-book-card-delete-icon"
              fill="currentColor"
              height="1em"
              role="button"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
